pinot-query-planner/src/main/java/org/apache/pinot/query/planner/nodes/RexExpression.java

```diff
@@ -0,0 +1,157 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.nodes;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.util.NlsString;
+import org.apache.pinot.query.planner.nodes.serde.ProtoProperties;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+
+/**
+ * {@code RexExpression} is the serializable format of the {@link RexNode}.
+ */
+public abstract class RexExpression {
+  @ProtoProperties
+  protected SqlKind _sqlKind;
+  @ProtoProperties
+  protected RelDataType _dataType;
+
+  public SqlKind getKind() {
+    return _sqlKind;
+  }
+
+  public RelDataType getDataType() {
+    return _dataType;
+  }
+
+  public static RexExpression toRexExpression(RexNode rexNode) {
+    if (rexNode instanceof RexInputRef) {
+      return new RexExpression.InputRef(((RexInputRef) rexNode).getIndex());
+    } else if (rexNode instanceof RexLiteral) {
+      RexLiteral rexLiteral = ((RexLiteral) rexNode);
+      return new RexExpression.Literal(rexLiteral.getType(), rexLiteral.getTypeName(), rexLiteral.getValue());
+    } else if (rexNode instanceof RexCall) {
+      RexCall rexCall = (RexCall) rexNode;
+      List<RexExpression> operands = rexCall.getOperands().stream().map(RexExpression::toRexExpression)
+          .collect(Collectors.toList());
+      return new RexExpression.FunctionCall(rexCall.getKind(), rexCall.getType(), rexCall.getOperator().getName(),
+          operands);
+    } else {
+      throw new IllegalArgumentException("Unsupported RexNode type with SqlKind: " + rexNode.getKind());
+    }
+  }
+
+  private static Comparable convertLiteral(Comparable value, SqlTypeName sqlTypeName, RelDataType dataType) {
+    switch (sqlTypeName) {
+      case BOOLEAN:
+        return (boolean) value;
+      case DECIMAL:
+        switch (dataType.getSqlTypeName()) {
+          case INTEGER:
+            return ((BigDecimal) value).intValue();
+          case BIGINT:
+            return ((BigDecimal) value).longValue();
+          case FLOAT:
+            return ((BigDecimal) value).floatValue();
+          case DOUBLE:
+          default:
+            return ((BigDecimal) value).doubleValue();
+        }
+      case CHAR:
+        switch (dataType.getSqlTypeName()) {
+          case VARCHAR:
+            return ((NlsString) value).getValue();
+          default:
+            return value;
+        }
+      default:
+        return value;
+    }
+  }
+
+  public static class InputRef extends RexExpression {
```

Comment:
I think this is the only one that is currently not modeled as part of our existing thrift `Expression`. It already supports Identifier, Literal and Function. So, can we not reuse `Literal` and `Function` below instead of modeling the same as `RexExpression` because eventually `RexExpression` has to be converted to thrift Expression much like SqlNode to PinotQuery ?